### PR TITLE
Support @RunWith for JUnit 4

### PR DESCRIPTION
**Pull request by jdneo**
This PR is used to add JUnit 4 `@RunWith` support #272 
  - [x] Parsing the result from the JUnit output instead of from the LS searching result.
  - [x] Search the location when the range information is not available for a test item in the result page.
  - [x] Search `@RunWith` annotation in `FrameworkSearcher`
  - [x] Guess the file path if the URI is undefined when storing the results in the cache.

Items need to be further addressed:
  - [ ] Remove duplicated suites when running tests from the workspace root
  - [ ] Show status icon on `@RunWith` annotated class.

### CODE REVIEW

File: src/protocols.ts [hunk 1]
```diff
@@ -3,15 +3,17 @@
...
     displayName: string;
     fullName: string;
     uri: string;
     range: Range;
 }
 
-export interface ITestItem extends ITestItemBase {
-    children: ITestItem[];
+export interface ITestItem extends ILocation {
+    displayName: string;
```
Comment (Eskibear) [inline src/protocols.ts]: displayName and fullName already defined in `ILocation `

File: src/runners/baseRunner/BaseRunnerResultAnalyzer.ts [hunk 2]
```diff
@@ -41,35 +42,61 @@ export abstract class BaseRunnerResultAnalyzer {
 
     public feedBack(): ITestResult[] {
         const result: ITestResult[] = [];
+        const itemMap: Map<string, ITestItem> = new Map<string, ITestItem>();
         for (const test of this.tests) {
-            this.processTestItemRecursively(test, result);
+            this.flatTestItems(test, itemMap);
         }
+        this.parseResults(result, itemMap);
+        this.addSkippedTests(result, itemMap);
```
Comment (Eskibear) [inline src/runners/baseRunner/BaseRunnerResultAnalyzer.ts]: The correctness of `parseResults` and `addSkippedTests` are strongly related to the execution order, as they share the state of itemMap.

File: src/runners/baseRunner/BaseRunnerResultAnalyzer.ts [hunk 3]
```diff
@@ -79,8 +106,4 @@ export abstract class BaseRunnerResultAnalyzer {
             .replace(/\\t/gm, '\t')
             .replace(/\\b/gm, '\b');
     }
-
-    protected get outputRegex(): RegExp {
-        return this.regex;
-    }
-}
+}
```
Comment (Eskibear) [inline src/runners/baseRunner/BaseRunnerResultAnalyzer.ts]: missing newline

File: src/runners/baseRunner/BaseRunnerResultAnalyzer.ts [hunk 4]
```diff
@@ -41,35 +42,61 @@ export abstract class BaseRunnerResultAnalyzer {
...
-        return {
-            displayName: test.displayName,
-            fullName: test.fullName,
-            uri: Uri.parse(test.uri).toString(),
-            range: test.range,
-            result: testResultDetails,
-        };
+    protected addSkippedTests(result: ITestResult[], itemMap: Map<string, ITestItem>): void {
+        for (const item of itemMap.values()) {
+            result.push(Object.assign(defaultResult, {
```
Comment (Eskibear) [inline src/runners/baseRunner/BaseRunnerResultAnalyzer.ts]: value of `defaultResult` is changed on every execution of the line. 

Consider `Object.assign({}, defaultResult, { ... })` ?

File: src/runners/baseRunner/BaseRunnerResultAnalyzer.ts [hunk 5]
```diff
@@ -41,35 +42,61 @@ export abstract class BaseRunnerResultAnalyzer {
...
+            map.set(item.fullName, item);
+        } else if (item.children) {
+            item.children.forEach((child: ITestItem) => this.flatTestItems(child, map));
         }
     }
 
-    protected abstract processData(data: string): void;
+    protected parseResults(resultArray: ITestResult[], itemMap: Map<string, ITestItem>): void {
+        for (const [key, value] of this.testResults.entries()) {
+            let result: ITestResult = Object.assign(_.cloneDeep(defaultResult), {
```
Comment (Eskibear) [inline src/runners/baseRunner/BaseRunnerResultAnalyzer.ts]: Why use clone**Deep** here?

Comment (jdneo) [reply src/runners/baseRunner/BaseRunnerResultAnalyzer.ts]: Will change to `Object.assign({}, defaultResult, { ... })`

File: src/testReportProvider.ts [hunk 6]
```diff
@@ -43,17 +44,29 @@ class TestReportProvider implements Disposable {
...
+                        if (items.length === 1) {
+                            return openTextDocument(Uri.parse(items[0].uri), items[0].range);
+                        } else if (items.length > 1) {
+                            const pick: ILocationQuickPick | undefined = await window.showQuickPick(items.map((item: ILocation) => {
+                                return { label: message.fullName, detail: Uri.parse(item.uri).fsPath, location: item };
+                            }), { placeHolder: 'Select the file you want to navigate to' });
+                            if (pick) {
+                                return openTextDocument(Uri.parse(pick.location.uri), pick.location.range);
+                            }
+                        }
```
Comment (Eskibear) [inline src/testReportProvider.ts]: missing condition `items.length <= 0`

File: java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java [hunk 7]
```diff
@@ -219,6 +241,34 @@ public void acceptSearchMatch(SearchMatch match) throws CoreException {
         return searchResult;
     }
 
+    public static List<Location> searchLocation(List<Object> arguments, IProgressMonitor monitor) throws CoreException {
+        final List<Location> searchResult = new LinkedList<>();
+        if (arguments == null || arguments.size() == 0) {
+            throw new RuntimeException("Invalid aruguments to search the location.");
```
Comment (yaohaizh) [inline java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java]: IllegalArgumentException

File: src/protocols.ts [hunk 8]
```diff
@@ -3,15 +3,15 @@
...
-    fullName: string;
+export interface ILocation {
     uri: string;
     range: Range;
 }
 
-export interface ITestItem extends ITestItemBase {
-    children: ITestItem[];
+export interface ITestItem extends ILocation {
+    displayName: string;
```
Comment (yaohaizh) [inline src/protocols.ts]: Same as the java side code

**Pull request merged**